### PR TITLE
Add ValidatesContainer tests with installing release candidates(RCs)

### PR DESCRIPTION
**Pull request by AnandInguva**
Part 2 of #25668 

Adds a Jenkins suite that run `ValidatesContainer` suite with pre-released Beam's dependencies. This could help catch regressions/breaking changes sooner than later.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by damccorm** [comment]
Run Python RC Dataflow ValidatesContainer

**Comment by AnandInguva** [comment]
Looks like there are other failures due to this PR. I will solve them. Changing this PR to draft until then.

**Comment by AnandInguva** [comment]
For reference: Successful build: https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont_with_RC_PR/1/

**Comment by AnandInguva** [comment]
cc: @tvalentyn

**Comment by Abacn** [comment]
I see failed seedjob on this PR: https://ci-beam.apache.org/job/beam_SeedJob/11381/console with "ERROR: Couldn't find any revision to build." may need to rebase the latest master onto it to run, probably due to a force push on master earlier today?

File: sdks/python/container/Dockerfile [hunk 1]
```diff
@@ -37,11 +37,11 @@ RUN apt-get update && \
 ####
 # Install required packages for Beam Python SDK and common dependencies used by users.
 ####
-
+ARG install_release_candidates
```
Comment (tvalentyn) [inline sdks/python/container/Dockerfile:40]: this ARG has no knowledge of what you pass; you can call it `pip_extra_options`  for example.

Comment (tvalentyn) [reply sdks/python/container/Dockerfile:40]: does it need a default value?

Comment (AnandInguva) [reply sdks/python/container/Dockerfile:40]: yes. but i defined the default value to be null in common.gradle file. I can provide default here

File: sdks/python/container/Dockerfile [hunk 2]
```diff
@@ -70,9 +70,11 @@ RUN ccache --set-config=sloppiness=file_macro && ccache --set-config=hash_dir=fa
 ####
 # Install Apache Beam SDK. Use --no-deps and pip check to verify that all
 # necessary dependencies are specified in base_image_requirements.txt.
+# use ARG pip_extra_options to install release candidates(RCs) of Beam'
```
Comment (tvalentyn) [inline sdks/python/container/Dockerfile:73]: again, you are making an assumption that this file can't reason about. one may supply pip extra options for any purpose.

Comment (AnandInguva) [inline sdks/python/container/Dockerfile:73]:
```suggestion
# use ARG pip_extra_options to pass additional flags to the pip installation.'
```

Comment (AnandInguva) [reply sdks/python/container/Dockerfile]: yes. previously the purpose was to add only `--pre` but since we can add any flag to pip installation, i changed the wording.

**Comment by AnandInguva** [comment]
retest this please

**Comment by AnandInguva** [comment]
Run PythonRC PreCommit

**Comment by AnandInguva** [comment]
Run Python RC Dataflow ValidatesContainer

**Comment by AnandInguva** [comment]
Some more work needs to be done. Converting to draft until then

File: sdks/python/container/run_generate_requirements.sh [hunk 3]
```diff
@@ -61,8 +62,8 @@ pip install --upgrade pip setuptools wheel
 # Install dataframe deps to add have Dataframe support in released images.
 # Install test deps since some integration tests need dependencies,
 # such as pytest, installed in the runner environment.
-pip install --no-cache-dir $SDK_TARBALL[gcp,dataframe,test]
-pip install --no-cache-dir -r $PWD/sdks/python/container/base_image_requirements_manual.txt
+pip install --pre --no-cache-dir $SDK_TARBALL[gcp,dataframe,test]
```
Comment (AnandInguva) [inline sdks/python/container/run_generate_requirements.sh:65]: replace `--pre` with PIP_EXTRA_OPTIONS once testing is done

**Comment by AnandInguva** [comment]
Run Python Dataflow ValidatesContainer

**Comment by AnandInguva** [comment]
Run Python Dataflow ValidatesContainer

**Comment by AnandInguva** [comment]
Run Python Dataflow ValidatesContainer

File: sdks/python/container/run_generate_requirements.sh [hunk 4]
```diff
@@ -30,39 +30,46 @@
...
+if [[ $# -lt 2 ]]; then
   printf "Example usage: \n$> ./sdks/python/container/run_generate_requirements.sh 3.8 <sdk_tarball>"
-  printf "\n\twhere 3.8 is the Python major.minor version."
+  printf "\n\where 3.8 is the Python major.minor version."
   exit 1
 fi
 
 PY_VERSION=$1
 SDK_TARBALL=$2
+PIP_EXTRA_OPTIONS=$3
```
Comment (AnandInguva) [inline sdks/python/container/run_generate_requirements.sh:41]: todo: provide default value to be ""

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 5]
```diff
@@ -2739,7 +2739,7 @@ class BeamModulePlugin implements Plugin<Project> {
       // If none of them applied, version set here will be used as default value.
       // TODO(BEAM-12000): Move default value to Py3.9.
```
Comment (tvalentyn) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2740]: remove the todo

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 6]
```diff
@@ -2739,7 +2739,7 @@ class BeamModulePlugin implements Plugin<Project> {
...
       project.ext.pythonVersion = project.hasProperty('pythonVersion') ?
-          project.pythonVersion : '3.8'
+          project.pythonVersion : '3.10'
```
Comment (tvalentyn) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2742]: for my education - is this a driveby change or this was necessary?

File: sdks/python/container/run_generate_requirements.sh [hunk 7]
```diff
@@ -30,39 +30,46 @@
...
+source "$ENV_PATH"/bin/activate
 pip install --upgrade pip setuptools wheel
 
 # Install gcp extra deps since these deps are commonly used with Apache Beam.
 # Install dataframe deps to add have Dataframe support in released images.
 # Install test deps since some integration tests need dependencies,
 # such as pytest, installed in the runner environment.
-pip install --no-cache-dir $SDK_TARBALL[gcp,dataframe,test]
-pip install --no-cache-dir -r $PWD/sdks/python/container/base_image_requirements_manual.txt
+if [ -z "$PIP_EXTRA_OPTIONS" ]; then
```
Comment (tvalentyn) [inline sdks/python/container/run_generate_requirements.sh:65]: you don't need the branch. default value is empty.

File: sdks/python/test-suites/dataflow/common.gradle [hunk 8]
```diff
@@ -351,6 +351,10 @@ task chicagoTaxiExample {
 
 task validatesContainer() {
   def pyversion = "${project.ext.pythonVersion.replace('.', '')}"
+  if (project.hasProperty("pipExtraOptions")) {
```
Comment (tvalentyn) [inline sdks/python/test-suites/dataflow/common.gradle:354]: this is confusing when you read it outside of the context of this change.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 6 above)
Comment (AnandInguva) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2742]: Drive by change. for my own debugging purpose on my laptop

File: sdks/python/test-suites/dataflow/common.gradle (hunk 8 above)
Comment (tvalentyn) [reply sdks/python/test-suites/dataflow/common.gradle:354]: pipExtraOptions in this context is a very generic flag and its unclear how it ties to regenerating requirements.

Comment (tvalentyn) [reply sdks/python/test-suites/dataflow/common.gradle:354]: in the context  `validatesContainer()` task it makes sense to use a more specific flag name (for example `regenerateRequiremetns`), while in the context of Dockerfile (where we run pip) it does make sense to use a generic flag name (`pipExtraOptions`).

File: sdks/python/container/run_generate_requirements.sh (hunk 7 above)
Comment (AnandInguva) [reply sdks/python/container/run_generate_requirements.sh:65]: Without the branch, I get this error
```
 pip install '' --no-cache-dir '/Users/anandinguva/projects/beam/sdks/python/build/apache-beam.tar.gz[gcp,dataframe,test]'
ERROR: Invalid requirement: ''
```

File: sdks/python/test-suites/dataflow/common.gradle (hunk 8 above)
Comment (AnandInguva) [reply sdks/python/test-suites/dataflow/common.gradle:354]: Sg. i will refactor the names

File: sdks/python/container/run_generate_requirements.sh (hunk 7 above)
Comment (AnandInguva) [reply sdks/python/container/run_generate_requirements.sh:65]: I can remove branch by using `${PIP_EXTRA_OPTIONS:+"$PIP_EXTRA_OPTIONS"} `. let me try

**Comment by AnandInguva** [comment]
Run Python RC Dataflow ValidatesContainer

**Comment by AnandInguva** [comment]
Run Python Dataflow ValidatesContainer

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 9]
```diff
@@ -2737,9 +2737,8 @@ class BeamModulePlugin implements Plugin<Project> {
       // Python interpreter version for virtualenv setup and test run. This value can be
       // set from commandline with -PpythonVersion, or in build script of certain project.
       // If none of them applied, version set here will be used as default value.
-      // TODO(BEAM-12000): Move default value to Py3.9.
       project.ext.pythonVersion = project.hasProperty('pythonVersion') ?
-          project.pythonVersion : '3.8'
+          project.pythonVersion : '3.11'
```
Comment (AnandInguva) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2741]: Looks like this change is non-trivial and need some changes. I will make a separate PR for this

**Comment by AnandInguva** [comment]
Run Python Dataflow ValidatesContainer

File: sdks/python/container/common.gradle [hunk 10]
```diff
@@ -32,13 +32,14 @@ dependencies {
 
 def generatePythonRequirements = tasks.register("generatePythonRequirements") {
   dependsOn ':sdks:python:sdist'
+  def enableRCInstallation = project.hasProperty("testRCDependencies") ? "--pre" : ""
```
Comment (tvalentyn) [inline sdks/python/container/common.gradle:35]:
```suggestion
  def pipExtraOptions = project.hasProperty("testRCDependencies") ? "--pre" : ""
```

File: sdks/python/container/common.gradle [hunk 11]
```diff
@@ -32,13 +32,14 @@ dependencies {
...
   def runScriptsPath = "${rootDir}/sdks/python/container/run_generate_requirements.sh"
   doLast {
     exec {
       executable 'sh'
       args '-c', "cd ${rootDir} && ${runScriptsPath} " +
               "${project.ext.pythonVersion} " +
-              "${files(configurations.sdkSourceTarball.files).singleFile}"
+              "${files(configurations.sdkSourceTarball.files).singleFile} " + "${enableRCInstallation}"
```
Comment (tvalentyn) [inline sdks/python/container/common.gradle:42]:
```suggestion
              "${files(configurations.sdkSourceTarball.files).singleFile} " + "${pipExtraOptions}"
```

**Review by tvalentyn — APPROVED** [review]

**Comment by AnandInguva** [comment]
@tvalentyn Can we merge this?

**Review by tvalentyn — APPROVED** [review]

**Comment by tvalentyn** [comment]
nice. Thanks!
